scd/p11token.c

```diff
@@ -0,0 +1,623 @@
+	// retrieve buffer size
+	P11_CHECK_RV_GOTO(token->ctx->C_Encrypt(*token->sh, plain_key, plain_key_len, NULL,
+						wrapped_key_len),
+			  error);
+	ptr = (unsigned char *)mem_alloc0(*wrapped_key_len);
+	IF_TRUE_GOTO_DEBUG(NULL == ptr, error);
+	P11_CHECK_RV_GOTO(token->ctx->C_Encrypt(*token->sh, plain_key, plain_key_len, ptr,
+						wrapped_key_len),
+			  error);
+	*wrapped_key = ptr;
+	return 0;
+error:
+	if (ptr) {
+		mem_free0(ptr);
+	}
+	return -1;
+}
+
+int
+p11token_unwrap_key(p11token_t *token, unsigned char *wrapped_key, size_t wrapped_key_len,
+		    unsigned char **plain_key, unsigned long *plain_key_len)
+{
+	ASSERT(token);
+	ASSERT(wrapped_key);
+	ASSERT(plain_key);
+	ASSERT(plain_key_len);
+
+	unsigned char *ptr = NULL;
+
+	if (p11token_is_locked(token)) {
+		ERROR("p11token_wrap_key: token is locked");
+		goto error;
+	}
+
+	// get wrapping key handle
+	// currently there is only one, therefore we can keep the template simple
+	ck_key_type_t aes_key_type = CKK_AES;
+	struct ck_attribute search_template[] = {
+		{ CKA_KEY_TYPE, &aes_key_type, sizeof(aes_key_type) },
+	};
+	P11_CHECK_RV_GOTO(token->ctx->C_FindObjectsInit(*token->sh, search_template,
+							ELEMENTSOF(search_template)),
+			  error);
+	ck_object_handle_t h_key;
+	unsigned long num_objects_found;
+	P11_CHECK_RV_GOTO(token->ctx->C_FindObjects(*token->sh, &h_key, 1, &num_objects_found),
+			  error);
+	P11_CHECK_RV_GOTO(token->ctx->C_FindObjectsFinal(*token->sh), error);
+	if (0 == num_objects_found) {
+		goto error;
+	}
+	// unwrap key
+	/**
+	 * static default IV as defined in RFC 3394
+	 * TODO: investigate whether a random IV which is passed along with the
+	 * 			wrapped key is possible and desirable
+	 */
+	unsigned char iv[] = { 0xA6, 0xA6, 0xA6, 0xA6, 0xA6, 0xA6, 0xA6, 0xA6,
+			       0xA6, 0xA6, 0xA6, 0xA6, 0xA6, 0xA6, 0xA6, 0xA6 };
+	struct ck_mechanism wrap_mechanism = {
+		CKM_AES_CBC_PAD,
+		iv,
+		ELEMENTSOF(iv),
+	};
+	P11_CHECK_RV_GOTO(token->ctx->C_DecryptInit(*token->sh, &wrap_mechanism, h_key), error);
+	P11_CHECK_RV_GOTO(token->ctx->C_Decrypt(*token->sh, wrapped_key, wrapped_key_len, NULL,
+						plain_key_len),
+			  error);
+	ptr = (unsigned char *)mem_alloc0(*plain_key_len);
+	IF_TRUE_GOTO_DEBUG(NULL == ptr, error);
+	P11_CHECK_RV_GOTO(token->ctx->C_Decrypt(*token->sh, wrapped_key, wrapped_key_len, ptr,
+						plain_key_len),
+			  error);
+	*plain_key = ptr;
+
+	return 0;
+error:
+	if (ptr) {
+		mem_free0(ptr);
+	}
+	return -1;
+}
+
+int
+p11token_change_pin(p11token_t *token, const char *old_pin, const char *new_pin)
+{
+	ASSERT(token);
+	ASSERT(old_pin);
+	ASSERT(new_pin);
+
+	if (-1 == p11token_unlock(token, old_pin)) {
+		goto error;
+	}
+
+	P11_CHECK_RV_GOTO(token->ctx->C_SetPIN(*token->sh, (unsigned char *)old_pin,
+					       strlen(old_pin), (unsigned char *)new_pin,
+					       strlen(new_pin)),
+			  error);
+
+	return p11token_lock(token);
+error:
+	return -1;
+}
+
+// internal helper functions
```

Comment:
put those functions at the beginning of the c file where you now have the prototype definitions.
remove the prototype definitions and declare those functions static.